pyomo/opt/base/solvers.py

```diff
@@ -640,7 +640,6 @@ def solve(self, *args, **kwds):
                             self._default_variable_value
                         if self._load_solutions:
                             _model.load_solution(result.solution(0))
-                            result.solution.clear()
```

Comment:
@ghackebeil If you delete this line, the solution will never be removed from the results object, right? Maybe the ValueError you added below should fit inside this if statement before "result.solution.clear()"?

Reply:
My intention was to leave the solution in the results object either way. I think we are confusing users when they see a results object with no solution on it, but I agree that we should revisit this as part of a larger design discussion very soon.

I also did not want raise the ValueError before executing the next few lines below this one, where I clean the symbol map off of the model (placed there by the solver interface or writer).

This is all a mess write now (for both AML and Kernel), I'm just trying to provide a working alternative to base future discussions on.

I am okay with someone re-adding `result.solution.clear()` in an `else` statement after the block where I raise the exception if they have a strong opinion about this.

Reply:
Good points. Maybe it does make sense to leave the solutions on the results object. A larger design discussion is definitely warranted - which is really related to refactoring the solver interfaces. I am fine with this pull request as is for now.

Reply:
For the time being, I am OK with leaving the solution on the results object - as that has burned users in the past ("Why am I not getting the solution back from the solver?").  But @ghackebeil is right: this needs to be a larger design discussion as part of the writer/solver API rework that I hope will happen this summer: putting the solution into the results object is pretty slow and probably shouldn't be done unless a user specifically asks for it (it should be much more efficient to directly parse the solution into the model).  The "number of solutions" value in the results object should be the number returned from the solver and not the number stored in the results object.  Finally, we should re-think how the solutions are stored in the results object and use something more efficient than the current mechanism.

Reply:
I agree on all counts. I like what Michael is doing in the persistent interface where you can tell the solver to load the solution (possibly partially) from the most recent solve in a separate step. Perhaps this could be generalized for non-persistent interfaces as a way to avoid a temporary solutions object, while still allow one to delay loading until after statuses are check. Another idea would be to allow users to provide a callback to the solve method that indicates whether or not they want the solution loaded based on statuses that are placed on the results object.